nixos/modules/system/etc/make-etc.sh

```diff
@@ -1,45 +1,57 @@
-source $stdenv/setup
-
-mkdir -p $out/etc
-
-set -f
-sources_=($sources)
-targets_=($targets)
-modes_=($modes)
-users_=($users)
-groups_=($groups)
-set +f
-
-for ((i = 0; i < ${#targets_[@]}; i++)); do
-    source="${sources_[$i]}"
-    target="${targets_[$i]}"
-
-    if [[ "$source" =~ '*' ]]; then
+# Ensure the script fails properly
+set -euo pipefail
+
+# shellcheck disable=SC1091,SC2154
+source "$stdenv/setup"
+
+# shellcheck disable=SC2154
+mkdir -p "$out/etc"
+
+# This reads the JSON into a big array named 'entries'.
+# Essentially, the jq converts every entry into a \0-separated
+# string of the specified fields source, mode, ....
+# Then all these strings are \0-separated as well and fed into
+# readarray, resulting in a big list of blocks of the 5 fields.
+# \0-separating the strings ensures **any** character can be part
+# of any of the fields (since \0 is not a valid character in paths).
+# shellcheck disable=SC2154
+readarray -d $'\0' -t entries < <(jq -rj '[.[] | [.source, .target, .mode, .user, .group] | join("\u0000")] | join("\u0000")' < "${entriesPath}")
+
+ret=0
+for ((i = 0; i < ${#entries[@]}; i+=5)); do
+    src="${entries[$i]}"
```

Comment:
This still seems a little hacky to me. How about doing this instead, I used this myself before too when ingesting json output:
```
jq '.[]' -c "$entriesPath" | while read -r line; do
  src=$(jq -r '.source' <<< "$line")
  target=$(jq -r '.target' <<< "$line")
  # ...
```

The `-c` makes sure that each list entry is output on a separate line as JSON

Reply:
I had that idea as well but it's horribly slow in practice. It took like 10 seconds for my local system

Reply:
Oh damn, yeah I guess it's a whole bunch of jq calls..

Possibly `@sh` could be used in some way to make this a bit nicer

Reply:
I tried that. @ajs124 helped me like 2 hours to get it to work using `@sh` but we were unable to. The main issue is that you get \n-separated values which you put into something like `while read -r source target …; do`. This however breaks when your value contains \n since it's a literal newline and that results in the loop running twice for half an entry each

Reply:
Which could potentially be fixed with \0 separated output, but jq doesn't support that in any released version.

Reply:
Alternative idea: What if we sidestep JSON and jq completely by just generating the shell script in Nix itself. So instead of a shell for loop, we have a `lib.concatMapStringsSep` or so. This will generate a fairly big script, but I don't see a problem with that.

Reply:
Won't this cause the same problems as https://github.com/svanderburg/node2nix/issues/46#issuecomment-801378717?

Reply:
I'd think the scales we are dealing with are a bunch smaller.

Reply:
On my systems the `etc` derivation is typically around 20k with 150-200 entries for the sources. So yes, that's a lot smaller.